theano/gof/fg.py

```diff
@@ -478,8 +481,14 @@ def find_path_to(output_var, input_var):
 
         for node in new_nodes:
             assert node not in self.apply_nodes
-            self.__setup_node__(node)
-            self.apply_nodes.add(node)
+            prevent_addition = False
+            for n in self._removed_nodes :
+                if node is n :
+                    prevent_addition = True
```

Comment:
Why do a loop here? This is pretty not efficient. You can just do:
prevent_addition = node in self._removed_nodes

this will be faster as it won't loop over the old nodes. It will use the set code that use hashing for this. So the cost will be mostly constant vs having a cost that is linear in the number of old nodes.